File: kotlin/wallet/src/main/kotlin/com/hopae/eudi/wallet/CredentialFilter.kt
package com.hopae.eudi.wallet

import com.hopae.eudi.wallet.spi.CredentialFormat

/** Local (non-network) predicate over stored credentials. For DCQL semantics use `credentials.match`. */
sealed interface CredentialFilter {
    fun matches(credential: Credential): Boolean

    data object All : CredentialFilter {
        override fun matches(credential: Credential): Boolean = true
    }

    data class ByVct(val vct: String) : CredentialFilter {
        override fun matches(credential: Credential): Boolean =
            (credential.format as? CredentialFormat.SdJwtVc)?.vct == vct
    }

    data class ByDocType(val docType: String) : CredentialFilter {
        override fun matches(credential: Credential): Boolean =
            (credential.format as? CredentialFormat.MsoMdoc)?.docType == docType
    }

    companion object {
        val all: CredentialFilter = All
        fun byVct(vct: String): CredentialFilter = ByVct(vct)
        fun byDocType(docType: String): CredentialFilter = ByDocType(docType)
    }
}

File: swift/Sources/Wallet/CredentialFilter.swift
import WalletAPI

/// Local (non-network) predicate over stored credentials. For DCQL semantics use `credentials.match`.
public enum CredentialFilter {
    case all
    case byVct(String)
    case byDocType(String)

    func matches(_ credential: Credential) -> Bool {
        switch self {
        case .all:
            return true
        case let .byVct(vct):
            if case let .sdJwtVc(v) = credential.format { return v == vct }
            return false
        case let .byDocType(docType):
            if case let .msoMdoc(d) = credential.format { return d == docType }
            return false
        }
    }
}
